edb/server/pgcon/rust_transport.py

```diff
@@ -0,0 +1,475 @@
+    def send(self, message: memoryview) -> None: ...
+    def upgrade(self) -> None: ...
+    def parameter(self, name: str, value: str) -> None: ...
+    def cancellation_key(self, pid: int, key: int) -> None: ...
+    def state_changed(self, state: int) -> None: ...
+    def auth(self, auth: int) -> None: ...
+
+
+StateChangeCallback = Callable[[ConnectionStateType], None]
+
+
+def _parse_tls_version(tls_version: str) -> ssl_module.TLSVersion:
+    if tls_version.startswith('SSL'):
+        raise ValueError(
+            f"Unsupported TLS version: {tls_version}"
+        )
+    try:
+        return ssl_module.TLSVersion[tls_version.replace('.', '_')]
+    except KeyError:
+        raise ValueError(
+            f"No such TLS version: {tls_version}"
+        )
+
+
+def _create_ssl(ssl_config: Dict[str, Any]):
+    sslmode = SSLMode.parse(ssl_config['sslmode'])
+    ssl = ssl_module.SSLContext(ssl_module.PROTOCOL_TLS_CLIENT)
+    ssl.check_hostname = sslmode >= SSLMode.verify_full
+    if sslmode < SSLMode.require:
+        ssl.verify_mode = ssl_module.CERT_NONE
+    else:
+        if ssl_config['sslrootcert']:
+            ssl.load_verify_locations(ssl_config['sslrootcert'])
+            ssl.verify_mode = ssl_module.CERT_REQUIRED
+        else:
+            if sslmode == SSLMode.require:
+                ssl.verify_mode = ssl_module.CERT_NONE
+        if ssl_config['sslcrl']:
+            ssl.load_verify_locations(ssl_config['sslcrl'])
+            ssl.verify_flags |= ssl_module.VERIFY_CRL_CHECK_CHAIN
+    if ssl_config['sslkey'] and ssl_config['sslcert']:
+        ssl.load_cert_chain(ssl_config['sslcert'],
+                            ssl_config['sslkey'],
+                            ssl_config['sslpassword'] or '')
+    if ssl_config['ssl_max_protocol_version']:
+        ssl.maximum_version = _parse_tls_version(
+            ssl_config['ssl_max_protocol_version'])
+    if ssl_config['ssl_min_protocol_version']:
+        ssl.minimum_version = _parse_tls_version(
+            ssl_config['ssl_min_protocol_version'])
+    # OpenSSL 1.1.1 keylog file
+    if hasattr(ssl, 'keylog_filename'):
+        if ssl_config['keylog_filename']:
+            ssl.keylog_filename = ssl_config['keylog_filename']
+    return ssl
+
+
+class PGConnectionProtocol(asyncio.Protocol):
+    """A protocol that manages the initial connection and authentication process
+    for PostgreSQL.
+
+    This protocol acts as an intermediary between the raw socket connection and
+    the user's protocol.
+    """
+    def __init__(
+            self,
+            state: PyConnectionState,
+            protocol: asyncio.Protocol,
+            ready_future: asyncio.Future,
+            pg_state: PGState,
+        ):
+        self.state = state
+        self.pg_state = pg_state
+        self.protocol = protocol
+        self.ready_future = ready_future
+        self._writing_paused = False
+        self._ready = False
+
+    def data_received(self, data: bytes):
```

Comment:
There's also the `eof_received()`; though we didn't quite use it much previously, we may do so later, so let's forward the callback here too so that we don't have to think about it again.

Reply:
As this code shouldn't be long-lived I think we can skip this. I hope to remove as much of that calling Python code as possible.